### PR TITLE
Add resource type completion for explain

### PR DESCRIPTION
**Pull request by nak3**
**What this PR does / why we need it**:

This PR supports completion for resource type list of kubectl explain.
After this, resource list could be completed by `kubectl explain [TAB][TAB]`.

**Release note**:

```release-note
NONE
```

Fixes https://github.com/kubernetes/kubernetes/issues/63173

### CODE REVIEW

File: pkg/kubectl/cmd/explain.go [hunk 1]
```diff
@@ -62,13 +62,15 @@ func NewExplainOptions(streams genericclioptions.IOStreams) *ExplainOptions {
 // NewCmdExplain returns a cobra command for swagger docs
 func NewCmdExplain(f cmdutil.Factory, streams genericclioptions.IOStreams) *cobra.Command {
 	o := NewExplainOptions(streams)
+	validArgs := cmdutil.ValidArgList(f)
```
Comment (liggitt) [inline pkg/kubectl/cmd/explain.go]: I'm not sure this method of hardcoding resource names into commands is something we should be extending

**Comment by liggitt** [comment]
@kubernetes/sig-cli-pr-reviews

File: pkg/kubectl/cmd/explain.go (hunk 1 above)
Comment (liggitt) [reply pkg/kubectl/cmd/explain.go]: I agree, hardcoding that list isn't really correct. Using the `api-resources` command to see what resources the server you are talking to has is more correct (but doesn't work well for static help and autocomplete)

**Comment by nak3** [comment]
I'm not sure if I should open another fresh PR or not. But first approach
should be updated, so added an util to get resource name lists, and
use it for `validArgs` to complete `kubectl explain`.

File: pkg/kubectl/cmd/explain.go [hunk 2]
```diff
@@ -76,9 +76,10 @@ func NewCmdExplain(parent string, f cmdutil.Factory, streams genericclioptions.I
 	cmd := &cobra.Command{
 		Use: "explain RESOURCE",
 		DisableFlagsInUseLine: true,
-		Short:   i18n.T("Documentation of resources"),
-		Long:    explainLong + "\n\n" + cmdutil.SuggestApiResources(parent),
-		Example: explainExamples,
+		Short:     i18n.T("Documentation of resources"),
+		Long:      explainLong + "\n\n" + cmdutil.SuggestApiResources(parent),
+		Example:   explainExamples,
+		ValidArgs: cmdutil.AllResourceList(f),
```
Comment (liggitt) [inline pkg/kubectl/cmd/explain.go]: this contacting the server during command construction, which isn't appropriate (it means this runs every time you invoke kubectl, regardless of whether you are running `explain`, and if a server is unavailable, can hang running any command or even displaying help)

Comment (liggitt) [reply pkg/kubectl/cmd/explain.go]: opened https://github.com/kubernetes/kubernetes/pull/63421 as an alternative that does the lookup when needed for completion

**Pull request closed without merging**